src/sshagent/OpenSSHKey.cpp

```diff
@@ -0,0 +1,480 @@
+}
+
+const QString OpenSSHKey::cipherName() const
+{
+    return m_cipherName;
+}
+
+const QString OpenSSHKey::type() const
+{
+    return m_type;
+}
+
+int OpenSSHKey::keyLength() const
+{
+    if (m_type == "ssh-dss" && m_publicData.length() == 4) {
+        return (m_publicData[0].length() - 1) * 8;
+    } else if (m_type == "ssh-rsa" && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 8;
+    } else if (m_type.startsWith("ecdsa-sha2-") && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 4;
+    } else if (m_type == "ssh-ed25519" && m_publicData.length() == 1) {
+        return m_publicData[0].length() * 8;
+    }
+
+    return 0;
+}
+
+const QString OpenSSHKey::fingerprint() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    QByteArray rawHash = QCryptographicHash::hash(publicKey, QCryptographicHash::Sha256);
+
+    return "SHA256:" + QString::fromLatin1(rawHash.toBase64(QByteArray::OmitTrailingEquals));
+}
+
+const QString OpenSSHKey::comment() const
+{
+    return m_comment;
+}
+
+const QString OpenSSHKey::publicKey() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    return m_type + " " + QString::fromLatin1(publicKey.toBase64()) + " " + m_comment;
+}
+
+const QString OpenSSHKey::errorString() const
+{
+    return m_error;
+}
+
+void OpenSSHKey::setType(const QString& type)
+{
+    m_type = type;
+}
+
+void OpenSSHKey::setPublicData(const QList<QByteArray>& data)
+{
+    m_publicData = data;
+}
+
+void OpenSSHKey::setPrivateData(const QList<QByteArray>& data)
+{
+    m_privateData = data;
+}
+
+void OpenSSHKey::setComment(const QString& comment)
+{
+    m_comment = comment;
+}
+
+void OpenSSHKey::clearPrivate()
+{
+    m_rawPrivateData.clear();
+    m_privateData.clear();
+}
+
+bool OpenSSHKey::parsePEM(const QByteArray& in, QByteArray& out)
+{
+    QString pem = QString::fromLatin1(in);
+    QStringList rows = pem.split(QRegExp("[\r\n]"), QString::SkipEmptyParts);
+
+    if (rows.length() < 3) {
+        m_error = tr("Invalid key file, expecting an OpenSSH key");
+        return false;
+    }
+
+    QString begin = rows.first();
+    QString end = rows.last();
+
+    QRegExp beginEx("-----BEGIN (.+)-----");
+    QRegExp endEx("-----END (.+)-----");
+
+    if (!beginEx.exactMatch(begin) || !endEx.exactMatch(end)) {
+        m_error = tr("Invalid key file, expecting an OpenSSH key");
+        return false;
+    }
+
+    if (beginEx.cap(1) != endEx.cap(1)) {
+        m_error = tr("PEM boundary mismatch");
+        return false;
+    }
+
+    if (beginEx.cap(1) != "OPENSSH PRIVATE KEY") {
+        m_error = tr("This is not an OpenSSH key, only modern keys are supported");
```

Comment:
This error is obtuse. What is meant by "modern keys"?? None of my keys loaded up and I have absolutely no idea how to fix the problem. Why isn't ```RSA PRIVATE KEY``` accepted?

Reply:
I don't know what to call the new type keys. Suggestions welcome.

Reply:
After researching it they are elliptic curve generated keys (ie, ed25519). Why don't you allow loading RSA keys? RSA 4096-bits is perfectly secure...

Reply:
It means the new "OpenSSH file format", it supports all key types and conversions are possible as stated in OP with instructions.

Reply:
Gotcha, I guess I should read your description first. Either way the link to the documentation (could even be within the GUI layout, not the error string) needs to be settled for the lay user. We will get SO MANY bug reports as it is implemented now.

Reply:
Would it be a link to the wiki? Is there any other reasonable way to document things like this?

Reply:
Sadly, we miss a lot of documentation, it's an open issue #696